livre/core/pipeline/FilterInput.h

```diff
@@ -0,0 +1,160 @@
+
+    /**
+     * Waits for a given port name or for all the input ports if
+     * portname is empty.
+     * @param portName is the name of the port. Empty for all ports.
+     * @throw std::runtime_error if the port with the port name
+     * is not available.
+     */
+    void wait( const std::string& portName = "" ) const;
+
+    /**
+     * Waits for a any ready input to a given port ( The port
+     * can have many inputs )
+     * @param portName is the name of the port.
+     * @return true there are still waiting inputs.
+     * @throw std::runtime_error if the port with the port name
+     * is not available.
+     */
+    bool waitForAny( const std::string& portName ) const;
+
+    /**
+     * Waits for a any input to be ready.
+     * @param readyPorts The ready ports are listed into readyPorts array.
+     * @return true there are still waiting inputs.
+     */
+    bool waitForAny( lunchbox::Strings& readyPorts ) const;
+
+    /**
+     * Queries given port name or for all the input ports if
+     * port is ready.
+     * @param portName is the name of the port. Empty for all ports.
+     * @return true if port/ports are ready.
+     * is not available.
+     * @throw std::runtime_error if the port with the port name
+     */
+    bool isReady( const std::string& portName = "" ) const;
+
+    /**
+     * @return the list of port names.
+     */
+    lunchbox::Strings getPortNames() const;
+
+    /**
+     * @return true if there are no ports.
+     */
+    bool isEmpty() const;
+
+    /**
+     * Gets the count of connections to the given input port.
+     * @param portName is the name of the port.
+     * @return the number of inputs.
+     * @throw std::runtime_error if the port with the port name
+     * is not available.
+     */
+    size_t getInputSize( const std::string& portName ) const;
+
+    /**
+     * @param portName is the name of the port.
+     * @return the input port
+     * @throw std::runtime_error if the port with the port name
+     * is not available.
+     */
+    InputPortPtr getPort( const std::string& portName );
+
+    /**
+     * Adds a new input port according to the given port info.
+     * @param portInfo The port information
+     * @return the generated code.
+     * @throw std::runtime_error if the port with the port name
+     * is already registered.
+     */
+    InputPortPtr addPort( const PortInfo& portInfo );
+
+    /**
+     * Clears the ports.
+     */
+    void clear();
+
+    /**
+     * Connect the given pipe filters with the given port name. Both filters
+     * should have the port name.
+     * @param src is the source pipe filter.
+     * @param dst is the destination pipe filter.
+     * @param portName connection port name.
+     * @return true if connection is successful.
+     */
+    friend bool connectFilters( PipeFilterPtr src,
+                                PipeFilterPtr dst,
+                                const std::string& portName );
```

Comment:
Yes, it access the pimpl. It felt more natural to have a free function for this as well.